edgedb-client/src/client.rs

```diff
@@ -142,15 +149,62 @@ impl Connection {
         return &self.version
     }
     pub async fn passive_wait<T>(&mut self) -> T {
-        let mut buf = [0u8; 1];
-        self.input.read(&mut buf[..]).await.ok();
+        let (_, mut reader, _) = self.split();
+        reader.passive_wait().await.ok();
         // any erroneous or successful read (even 0) means need reconnect
-        self.dirty = true;
+        self.state = State::Dirty;
         pending::<()>().await;
         unreachable!();
     }
+    async fn do_pings(&mut self) -> Result<(), Error> {
+        use async_std::io;
+
+        let ivl = self.ping_interval;
+        let (mut writer, mut reader, state) = self.split();
+
+        while *state == State::Normal {
+
+            *state = State::Dirty;
+            writer.send_messages(&[ClientMessage::Sync]).await?;
```

Comment:
Shouldn't we wait first, then send `Sync`? I think we could reuse the `State::Normal { last_ping: Instant }` to keep the waiting rhythm across `background_pings()` calls:

```patch
diff --git a/edgedb-client/src/builder.rs b/edgedb-client/src/builder.rs
index 03ac040..b1ce375 100644
--- a/edgedb-client/src/builder.rs
+++ b/edgedb-client/src/builder.rs
@@ -1049,14 +1049,14 @@ impl Builder {
         let mut version = ProtocolVersion::current();
         let (input, output) = stream.split();
         let mut conn = Connection {
             ping_interval: Duration::from_secs(50),
             input,
             output,
             input_buf: BytesMut::with_capacity(8192),
             output_buf: BytesMut::with_capacity(8192),
             params: TypeMap::custom(),
             transaction_state: TransactionState::NotInTransaction,
-            state: State::Normal,
+            state: State::Normal { last_pong: Instant::now() },
             version: version.clone(),
         };
         let mut seq = conn.start_sequence().await?;
@@ -1145,6 +1145,7 @@ impl Builder {
         }
         conn.version = version;
         conn.params = server_params;
+        conn.state = State::Normal { last_pong: Instant::now() };
         Ok(conn)
     }
 }
diff --git a/edgedb-client/src/client.rs b/edgedb-client/src/client.rs
index 56c98c4..a022c76 100644
--- a/edgedb-client/src/client.rs
+++ b/edgedb-client/src/client.rs
@@ -3,7 +3,7 @@
 use std::collections::HashMap;
 use std::fmt;
 use std::str;
-use std::time::Duration;
+use std::time::{Duration, Instant};
 
 use async_std::prelude::StreamExt;
 use async_std::future::{timeout, pending};
@@ -38,7 +38,7 @@ use crate::server_params::ServerParam;
 
 #[derive(Debug, PartialEq, Eq, Clone, Copy)]
 pub(crate) enum State {
-    Normal,
+    Normal { last_pong: Instant },
     Dirty,
     AwaitingPing,
 }
@@ -140,7 +140,7 @@ impl<'a> Sequence<'a> {
 
     pub fn end_clean(&mut self) {
         self.active = false;
-        *self.state = State::Normal;
+        *self.state = State::Normal { last_pong: Instant::now() };
     }
 }
 
@@ -162,22 +162,29 @@ impl Connection {
         let ivl = self.ping_interval;
         let (mut writer, mut reader, state) = self.split();
 
-        while *state == State::Normal {
-
-            *state = State::Dirty;
-            writer.send_messages(&[ClientMessage::Sync]).await?;
-            *state = State::AwaitingPing;
-            reader.wait_ready().await?;
-            *state = State::Normal;
+        if *state == State::AwaitingPing {
+            if reader.wait_ready().await.is_err() {
+                *state = State::Dirty;
+            } else {
+                *state = State::Normal { last_pong: Instant::now() };
+            }
+        }
 
-            match io::timeout(ivl, reader.passive_wait()).await {
-                Err(e) if e.kind() == io::ErrorKind::TimedOut => continue,
+        while let State::Normal {last_pong} = *state {
+            match io::timeout(ivl - (Instant::now() - last_pong), reader.passive_wait()).await {
+                Err(e) if e.kind() == io::ErrorKind::TimedOut => (),
                 Err(e) => {
                     *state = State::Dirty;
                     return Err(ClientConnectionError::with_source(e))?;
                 }
                 Ok(_) => unreachable!(),
             }
+
+            *state = State::Dirty;
+            writer.send_messages(&[ClientMessage::Sync]).await?;
+            *state = State::AwaitingPing;
+            reader.wait_ready().await?;
+            *state = State::Normal { last_pong: Instant::now() };
         }
         Ok(())
     }
@@ -199,12 +206,12 @@ impl Connection {
             if reader.wait_ready().await.is_err() {
                 self.state = State::Dirty;
             } else {
-                self.state = State::Normal;
+                self.state = State::Normal { last_pong: Instant::now() };
             }
         }
     }
     pub fn is_consistent(&self) -> bool {
-        self.state == State::Normal
+        matches!(self.state, State::Normal { last_pong: _ })
     }
     pub async fn terminate(mut self) -> Result<(), Error> {
         let mut seq = self.start_sequence().await?;
@@ -220,7 +227,7 @@ impl Connection {
         -> Result<Sequence<'x>, Error>
     {
         let (writer, reader, state) = self.split();
-        if *state != State::Normal {
+        if !matches!(*state, State::Normal { last_pong: _ }) {
             return Err(ClientInconsistentError::with_message(
                 "Connection is inconsistent state. Please reconnect."));
         }
```

And in CLI:

```patch
diff --git a/src/interactive.rs b/src/interactive.rs
index 5ac9e2f..3402740 100644
--- a/src/interactive.rs
+++ b/src/interactive.rs
@@ -573,6 +573,7 @@ async fn _interactive_main(options: &Options, state: &mut repl::State)
             prompt::Input::Value(_) => unreachable!(),
         };
         for item in ToDo::new(&inp) {
+            state.synchronize_ping().await;
             let result = match item {
                 ToDoItem::Backslash(text) => {
                     execute_backslash(state, text)
diff --git a/src/repl.rs b/src/repl.rs
index fadf489..6262051 100644
--- a/src/repl.rs
+++ b/src/repl.rs
@@ -169,10 +169,8 @@ impl State {
             ).await
             .context("cannot send to input thread")?;
         let result = if let Some(conn) = &mut self.connection {
-            let result = self.prompt.data.recv()
-                .race(conn.background_pings()).await;
-            conn.synchronize_ping().await;
-            result
+            self.prompt.data.recv()
+                .race(conn.background_pings()).await
         } else {
             self.prompt.data.recv().await
         };
@@ -181,6 +179,11 @@ impl State {
             Ok(x) => Ok(x),
         }
     }
+    pub async fn synchronize_ping(&mut self) {
+        if let Some(conn) = &mut self.connection {
+            conn.synchronize_ping().await;
+        }
+    }
     pub async fn input_mode(&mut self, value: InputMode) -> anyhow::Result<()>
     {
         self.input_mode = value;
```